src/scanspec/core.py

```diff
@@ -104,134 +102,106 @@ def calculate(self) -> int:
-    _tagged_unions[cls] = tagged_union
-    cls.__init_subclass__ = classmethod(partial(__init_subclass__, discriminator))
-    cls.__get_pydantic_core_schema__ = classmethod(
-        partial(__get_pydantic_core_schema__, tagged_union=tagged_union)
-    )
-    return cls
+    tagged_union = _TaggedUnion(super_cls, discriminator)
+    _tagged_unions[super_cls] = tagged_union
+
+    def add_subclass_to_union(subclass):
+        # Add a discriminator field to a subclass so it can
+        # be identified when deserializing
+        subclass.__annotations__ = {
+            **subclass.__annotations__,
+            discriminator: Literal[subclass.__name__],  # type: ignore
+        }
+        setattr(subclass, discriminator, Field(subclass.__name__, repr=False))  # type: ignore
 
+    def get_schema_of_union(cls, source_type: Any, handler: GetCoreSchemaHandler):
+        if cls is not super_cls:
+            tagged_union.add_member(cls)
+            return handler(cls)
+        # Rebuild any dataclass (including this one) that references this union
+        # Note that this has to be done after the creation of the dataclass so that
+        # previously created classes can refer to this newly created class
+        return tagged_union.schema(handler)
 
-T = TypeVar("T", type, Callable)
+    super_cls.__init_subclass__ = classmethod(add_subclass_to_union)  # type: ignore
+    super_cls.__get_pydantic_core_schema__ = classmethod(get_schema_of_union)  # type: ignore
+    return super_cls
 
 
-def deserialize_as(cls, obj):
-    return _tagged_unions[cls].type_adapter.validate_python(obj)
+T = TypeVar("T", type, Callable)
 
 
 def uses_tagged_union(cls_or_func: T) -> T:
     """
-    Decorator that processes the type hints of a class or function to detect and
-    register any tagged unions. If a tagged union is detected in the type hints,
-    it registers the class or function as a referrer to that tagged union.
-    Args:
-        cls_or_func (T): The class or function to be processed for tagged unions.
-    Returns:
-        T: The original class or function, unmodified.
+    T = TypeVar("T", type, Callable)
+        Decorator that processes the type hints of a class or function to detect and
+        register any tagged unions. If a tagged union is detected in the type hints,
+        it registers the class or function as a referrer to that tagged union.
+        Args:
+            cls_or_func (T): The class or function to be processed for tagged unions.
+        Returns:
+            T: The original class or function, unmodified.
     """
-    for k, v in get_type_hints(cls_or_func).items():
+    for v in get_type_hints(cls_or_func).values():
         tagged_union = _tagged_unions.get(get_origin(v) or v, None)
         if tagged_union:
-            tagged_union.add_referrer(cls_or_func, k)
+            tagged_union.add_reference(cls_or_func)
     return cls_or_func
 
 
+_tagged_unions: dict[type, _TaggedUnion] = {}
+
+
 class _TaggedUnion:
```

Comment:
I think it _should_ be possible to move some of this logic back into the `discriminated_union_of_subclasses` function now it is simpler. That might read better with it all in one place? Regardless that should probably be a future issue.